### PR TITLE
Check for valid subfolders when setting an array of allowed base URLs

### PR DESCRIPTION
**Pull request by bastianallgeier**
See: https://github.com/getkirby/kirby/issues/4191

### CODE REVIEW

**Comment by afbora** [comment]
I don't know how important but ending slash allowed url doesn't work:

````php
// doesn't work
return [
    'url' => [
        'http://localhost/test/kirby/develop/',
    ]
];

// works
return [
    'url' => [
        'http://localhost/test/kirby/develop',
    ]
];
````

**Comment by distantnative** [comment]
Given that users are very likely to do it both ways, I think it is rather important to support both

**Comment by distantnative** [comment]
@afbora are you able to create an [EnvironmentTest.php](https://github.com/getkirby/kirby/pull/4199#diff-3ca5a9e062bfedf0e51359b196dd81354fb9c88bb16a2f8e1134c7eac1972285) unit test that shows this fails?

File: src/Cms/Environment.php [hunk 1]
```diff
@@ -82,18 +77,36 @@ public function __construct(string $root, $allowed = null)
             // the current URL should be auto detected from a host allowlist
         } elseif (is_array($allowed) === true) {
             foreach ($allowed as $url) {
```
Comment (afbora) [inline src/Cms/Environment.php:79]:
```suggestion
            $hosts = [];
            foreach ($allowed as $url) {
```

**Comment by afbora** [comment]
@distantnative Yes, I can reproduce the exception with ending slash change in `EnvironmentTest::testAllowFromUrlsWithSubfolders()` test
https://github.com/getkirby/kirby/blob/fix/4191-subfolder-check/tests/Cms/Environment/EnvironmentTest.php#L154-L169

**Review by lukasbestle — COMMENTED** [review]
We could replace the foreach loop with something like this:

```php
$allowedUris = array_map(fn($url) => new Uri($url), $allowed);
$hosts       = array_map(fn($uri) => $uri->host(), $allowedUris);
```

Then we can do:

```php
$allowedStrings = array_map(fn($uri) => (string)$uri, $allowedUris);

if (in_array((string)$this->uri, $allowedStrings) {
    ...
}
```

**Comment by lukasbestle** [comment]
I suggest we make any further changes to the third and last PR. For review and discussion the separate PRs were really helpful, but if we now make changes here, we will have to rebase against a lot of merge conflicts. We can avoid that by committing to the third branch and then merging just that.

**Pull request closed without merging**